### PR TITLE
bpf: Remove bpf_netdev.o from previously used devices

### PR DESCRIPTION
**Pull request by brb**
This commit makes cilium-agent to remove `bpf_netdev.o` from devices which no longer suppose to have the program attached. This can happen when e.g. a user has specified a different device for NodePort via `--device` or they switched from the direct routing mode to the tunnel mode.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10087)

### CODE REVIEW

**Comment by brb** [comment]
test-me-please

File: bpf/init.sh [hunk 1]
```diff
@@ -561,6 +561,17 @@ else
...
 fi
 
+# Remove bpf_netdev.o from previously used devices
+for iface in $(ip -o -a l | awk '{print $2}' | cut -d: -f1 | grep -v cilium); do
+    [ "$iface" == "$NATIVE_DEV" ] && continue
+    if (tc filter show dev "$iface" ingress ; tc filter show dev "$iface" egress) | \
+		grep -q "bpf_netdev.o"; then
+
+        echo "Removing bpf_netdev.o from $iface"
+        tc qdisc del dev "$iface" clsact || true
```
Comment (borkmann) [inline bpf/init.sh]: I'm a bit worried about a corner case where e.g. on ingress we'd have bpf_netdev but on egress some other attached prog and then we'd remove the whole qdisc, which may not be intentional. Wasn't there a case where we attach bpf_network to every device? /cc @joestringer @jrfastab

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
CI hit the `RuntimeCassandra Tests policy allowing all actions` flake. Re-running.

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
CI provisioning failed. Re-running.

**Comment by brb** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
Any chance we can work towards unit testing of this kind of functionality? I guess for safety it'd need to run in its own namespace so not super straightforward.. but it'd be nice to be able to validate the corner cases of this.

Maybe this is an item for moving `bpf/init.sh` into Golang..

**Comment by brb** [comment]
test-me-please

**Review by borkmann — APPROVED** [review]

**Comment by brb** [comment]
@joestringer I was thinking about restarting cilium-agent with different configuration in CI, and checking whether progs were successfully removed from a prev device (added to my follow-up TODO list).

Having `bpf/init.sh` in Go would help us to unit test the behavior due to easier mocking in Go.

**Pull request merged**